src/rabin.js

```diff
@@ -0,0 +1,186 @@
+        throw "Error: Incorrect argument(s) to greatestCommonDivisor()";
+
+    a = BigIntMath.abs(a);
+    b = BigIntMath.abs(b);
+    if (b > a) {
+        let t = a;
+        a = b;
+        b = t;
+    }
+    while (b > 0) {
+        let t = b;
+        b = a % b;
+        a = t;
+    }
+    return a;
+}
+
+// Calculates: base^exponent % modulus
+function powerMod(base, exponent, modulus) {
+    if (modulus === 1n) return 0n;
+    let result = 1n;
+    base = base % modulus;
+    while (exponent > 0n) {
+        if (exponent % 2n === 1n) //odd number
+            result = (result * base) % modulus;
+        exponent = exponent >> 1n; //divide by 2
+        base = (base * base) % modulus;
+    }
+    return result;
+}
+
+function rabinHashBytes(bytes) {
+    hBytes = bsv.crypto.Hash.sha256(bytes);
+    let idx = hBytes.byteLength / 2;
+    let hl = bsv.crypto.Hash.sha256(hBytes.slice(0, idx));
+    let hr = bsv.crypto.Hash.sha256(hBytes.slice(idx, hBytes.byteLength));
+    return toBigIntLE(Buffer.concat([hl, hr]));
+}
+
+function calculateNextPrime(p) {
+    smallPrimesProduct = 3n * 5n * 7n * 11n * 13n * 17n * 19n * 23n * 29n;
+    while (greatestCommonDivisor(p, smallPrimesProduct) != 1) {
+        p = p + 4n;
+    }
+    if (powerMod(2n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(3n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(5n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    if (powerMod(17n, p - 1n, p) != 1n) {
+        return calculateNextPrime(p + 4n);
+    }
+    return p;
+}
+
+function getPrimeNumber(p) {
+    while (p % 4n != 3n) {
+        p = p + 1n;
+    }
+    return calculateNextPrime(p);
+}
+
+const paddingBuffer = Buffer.from('00', 'hex');
+
+function root(dataBuffer, p, q, nRabin) {
+    let sig, x, paddingByteCount = 0;
+    while (true) {
+        let h1 = rabinHashBytes(dataBuffer);
+        x = rabinHashBytes(dataBuffer) % nRabin;
+        sig = powerMod(p, q - 2n, q) * p * powerMod(x, (q + 1n) / 4n, q);
+        sig = (powerMod(q, p - 2n, p) * q * powerMod(x, (p + 1n) / 4n, p) + sig) % (nRabin);
+        if (((sig * sig) % nRabin) === x) {
+            break;
+        }
+        dataBuffer = Buffer.concat([dataBuffer, paddingBuffer]);
+        paddingByteCount++;
+    }
+    return {
+        "signature": decimalToHexString(sig),
+        "paddingByteCount": paddingByteCount
+    };
+}
+
+function calculateRabinValueFromKeyParts(p,q){
+    if((typeof(p) !== 'bigint' && isNaN(p)) || (typeof(q) !== 'bigint' && isNaN(q)))
+        throw("Error: Key parts should be numbers.")
+    return decimalToHexString(BigInt(p) * BigInt(q));
+}
+
+/**
+ * Generates Key p, q and nRabin values from a hexadecimal seed
+ * @param {String} seed hexadecimal string value
+ * @returns {JSON} {'p': String,'q': String, 'nRabin': String} Key p, q and nRabin values in hex
+ */
+function generateRabinKeyFromSeed(seed) {
```

Comment:
Let's remove the seed and use a PRNG instead.

Reply:
Got any specific PRNG in mind?

Reply:
Check what money button bsv library is using.